docs/Changelog.md

```diff
@@ -1010,6 +1010,249 @@ opset_import {
+              type {
+                tensor_type {
+                  elem_type: INT32
+                  shape {
+                  }
+                }
+              }
+            }
+            output {
+              name: "keepgoing_out"
+              type {
+                tensor_type {
+                  elem_type: BOOL
+                  shape {
+                  }
+                }
+              }
+            }
+            output {
+              name: "user_defined_vals"
+              type {
+                tensor_type {
+                  elem_type: INT32
+                  shape {
+                  }
+                }
+              }
+            }
+          }
+          type: GRAPH
+        }
+      }
+      name: "predict-net"
+  
+    == Sample equivalent C code ==
+  
+      {
+        /* User-defined code (enclosing scope) */
+        int a = 3, b = 6;
+        bool keepgoing = true; // Analogous to input cond
+        /* End user-defined code */
+  
+        /* Implicitly-defined code */
+        int i = 0; // implicitly defined trip count
+        const int max_trip_count = 10; // Analogous to input M
+        int user_defined_vals[]; // Imagine this is resizable
+        /* End implicitly-defined code */
+        do {
+          /* User-defined code (loop body) */
+          int my_local = a + b; // Reading values in the enclosing scope is fine
+          b = a - b; // writes fine if we specify b as a loop-carried dependency
+          keepgoing = my_local > b; // keepgoing is a loop-carried dependency
+          user_defined_vals[i] = b + b;
+          /* End user-defined code */
+  
+          i = i + 1; // implicitly defined trip count
+        } while (i < max_trip_count && keepgoing);
+        // my_local = 123; // Can't do this. my_local was defined in the the body
+      }
+  
+  There are several things of note in this code snippet:
+  
+  1) Reads from values in the enclosing scope (i.e. variable a here) are
+     implicitly allowed.
+  2) Writes to variables in the enclosing scope (i.e. the variables b and
+      keepgoing) MUST have the corresponding values appearing in the loop-carried
+      dependencies, both at the op input and body net input and output.
+  3) Values created in the body cannot be accessed in the enclosing scope.
+  
+  Note that the semantics of this op support "diagonal" or "wavefront" execution.
+  (See Step 3 here for an example:
+  https://devblogs.nvidia.com/optimizing-recurrent-neural-networks-cudnn-5/).
+  Frontends should emit multi-layer RNNs as a series of While operators (with
+  time being the inner looping dimension), with each successive layer consuming
+  the scan_outputs from the previous layer, possibly going through several
+  point-wise operators (e.g. dropout, residual connections, linear layer).
+  Backends can recognize this pattern and are permitted to schedule the execution
+  of the multi-layer network in a pipelined/"wavefront" fashion.
```

Comment:
In other words, ONNX evaluation order is not your plain old strict evaluation order :) I guess we should emphasize this in the overall ONNX docs too.